File: backend/app/ai/audit.py
import json
from datetime import datetime, timezone
from typing import Any

from app.logging_config import get_logger

logger = get_logger("ai.audit")


def log_llm_call(
    model: str,
    messages: list[dict[str, Any]],
    response_content: str,
    usage: dict[str, Any] | None,
    duration_ms: float,
    error: str | None = None,
) -> None:
    """Append an LLM call record to a daily JSONL audit log."""
    try:
        from app.config import settings
        audit_dir = settings.audit_dir
        audit_dir.mkdir(parents=True, exist_ok=True)

        date_str = datetime.now(timezone.utc).strftime("%Y-%m-%d")
        filepath = audit_dir / f"llm_{date_str}.jsonl"

        record = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "model": model,
            "messages": messages,
            "response": response_content,
            "usage": usage or {},
            "duration_ms": round(duration_ms, 1),
            "error": error,
        }

        with open(filepath, "a", encoding="utf-8") as f:
            f.write(json.dumps(record, ensure_ascii=False) + "\n")

        logger.debug("Audit logged to %s (%d chars response)", filepath, len(response_content))
    except Exception as e:
        logger.warning("Failed to write audit log: %s", e)

File: src-tauri/src/ai/audit.rs
use std::path::Path;

pub fn log_llm_call(
    data_dir: &Path,
    model: &str,
    messages: &[serde_json::Value],
    response_content: &str,
    usage: Option<&serde_json::Value>,
    duration_ms: f64,
    error: Option<&str>,
) {
    let audit_dir = data_dir.join("llm_audit");
    if let Err(e) = std::fs::create_dir_all(&audit_dir) {
        log::warn!("Failed to create audit dir: {e}");
        return;
    }

    let date_str = chrono::Utc::now().format("%Y-%m-%d");
    let filepath = audit_dir.join(format!("llm_{date_str}.jsonl"));

    let record = serde_json::json!({
        "timestamp": chrono::Utc::now().to_rfc3339(),
        "model": model,
        "messages": messages,
        "response": response_content,
        "usage": usage.cloned().unwrap_or(serde_json::json!({})),
        "duration_ms": (duration_ms * 10.0).round() / 10.0,
        "error": error,
    });

    if let Ok(json) = serde_json::to_string(&record) {
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&filepath)
        {
            use std::io::Write;
            let _ = writeln!(file, "{json}");
        }
    }
}
